modules/bq-log-alerting/logging/cloud_function/index.js

```diff
@@ -0,0 +1,165 @@
+  return {
+    type: `custom.googleapis.com/${name}`,
+    labels: labels
+  };
+}
+
+getResource = function(labels) {
+  return {
+    type: 'global',
+    labels: labels
+  };
+}
+
+getTimeseries = function(value) {
+  return [{
+    interval: {endTime: {seconds: Date.now() / 1000}},
+    value: value
+  }];
+}
+
+function createFindingObject (source_name, labels) {
+  const eventTime = new Date(labels.eventTimestamp.value).getTime();
+  const sourceProperties = {};
+  for (const key in labels) {
+    if (key !== 'eventTimestamp' &&
+        key !== 'resourceName' &&
+        key !== 'query') {
+      sourceProperties[key] = { stringValue: "" + labels[key] };
+    }
+  }
+  return {
+    parent: "" + source_name,
+    findingId: MD5("" + labels.query + labels.eventTimestamp + labels.callerIp + labels.principalEmail + labels.resourceName),
+    finding: {
+      state: 'ACTIVE',
+      // Resource the finding is associated with.  This is an
+      // example any resource identifier can be used.
+      resourceName: labels.resourceName,
+      // A free-form category.
+      category: "" + labels.query,
+      // The time associated with discovering the issue.
+      eventTime: {
+        seconds: Math.floor(eventTime / 1000),
+        nanos: (eventTime % 1000) * 1e6,
+      },
+      sourceProperties: sourceProperties
+    }
+  }
+}
+
+async function createFinding(cscc_client, source_name, labels) {
+  try {
+    finding = createFindingObject(source_name, labels)
+    if (dry_run) {
+      console.log('DRY_RUN: scc finding: ', JSON.stringify(finding));
+    } else {
+      const [newFinding] = await cscc_client.createFinding(finding);
+    }
+    return 1;
+  } catch (err) {
+    console.error("" + err);
+  }
+  return 0;
+}
+
+async function createFindingsFromResult(cscc_client, source_name, tableName, result) {
+  let count = 0;
+  for (var i = 0; i < result.length; i++) {
+    const res = result[i];
+    const labels = { query: tableName };
+    for (var j = 0; j < res.length; j++) {
+      const row = res[j];
+      for (var key in row) {
+        if (row.hasOwnProperty(key)) {
+          if (key !== 'receiveTimestamp') {
+            labels[key] = row[key];
+          } else {
+            // receiveTimestamp is returned as a complex type.
+            labels[key] = row[key]['value'];
+          }
+        }
+      }
+      if (labels.hasOwnProperty('resourceName')) {
+        count += await createFinding(cscc_client, source_name, labels);
+      }
+    }
+  }
+  return count;
+}
+
+exports.cronPubSub = async function (event, context, callback) {
+    const cscc_client = new SecurityCenterClient();
+    const pubsubMessage = event;
+    const payload = JSON.parse(Buffer.from(pubsubMessage.data, 'base64').toString()||'{}');
+    const quantity = parseInt(pubsubMessage.attributes ? (pubsubMessage.attributes.quantity ? pubsubMessage.attributes.quantity : 1) : (payload ? ( payload.quantity ? payload.quantity : 1) : 1));
```

Comment:
this ternary operation is a bit confusing, can we document or add a comment?

Reply:
code cleaned up for clarity